### ISSUE TITLE
 tf.keras model.fit calls slow with TPU distribute strategy

### ISSUE THREAD

**Issue by capilano** [tensorflow/tensorflow#30162]
<em>Please make sure that this is a bug. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Win 10
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): 
- TensorFlow version (use command below):1.14
- Python version:3,6
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version:
- GPU model and memory: N/A

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with: 1. TF 1.0: `python -c "import
tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"` 2. TF 2.0: `python -c
"import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**
TPU distribution strategy does not support model.fit_generator, and repeated model.fit calls result in a 50x slowdown presumably because it adds operations to graph. 

**Describe the expected behavior**

**Code to reproduce the issue**

resolver = tf.contrib.cluster_resolver.TPUClusterResolver(TPU_WORKER)
tf.contrib.distribute.initialize_tpu_system(resolver)
strategy =  tf.distribute.experimental.TPUStrategy(resolver)
with strategy.scope():
    model = ..... ## Your tf.keras model
    model.compile(loss = custom_loss,optimizer ='custom_optimizer)
 
for i in range(num_its):
      data,labels = = next(generator_fn()) 
      model.fit(data,labels)   

     

**Other info / logs**
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.

**Comment by achandraa** [comment]
In order to expedite the trouble-shooting process, please provide a code snippet to reproduce the issue reported here. Thanks!

**Comment by capilano** [comment]
I have just added a few lines to illustrate the issue. I have a very big workflow, so I cannot add the whole code here. The issue is that when model.fit is called in a loop(as shown above), the training slows down considerably.

**Comment by gadagashwini** [comment]
@capilano It is difficult to reproduce the issue. Will it be possible to provide us the minimal code snippet to reproduce the issue. So that we can reproduce the issue on our environment for faster resolution. Thanks!

**Comment by capilano** [comment]
`## Define your model ###
def network():
    base_model =ResNet50(weights=None, include_top=False)
    x = base_model.output
    x = GlobalAveragePooling2D()(x)
    x = Dense(1024, activation='relu')(x)
    predictions = Dense(10, activation='softmax')(x)
    model = Model(inputs=base_model.input, outputs=predictions)
   return model

### Initialize TPU system with TPU_WORKER as the TPU address ##
resolver = tf.contrib.cluster_resolver.TPUClusterResolver(TPU_WORKER)
tf.contrib.distribute.initialize_tpu_system(resolver)
strategy = tf.distribute.experimental.TPUStrategy(resolver)
with strategy.scope():
       model = network()
       model.compile(loss = 'categorical_crossentropy',optimizer ='adam') 
## Load data ##
x = np.ones((224,224,3))
n= np.rand.randint(0,9)
y = tf.one_hot(n,10)
data = []
labels = []
for i in range(1024):
      data.append(x)
     labels.append(y)
## Fit the model ##     (Dummy case just repeating the same data) 
for i in range(100000):  
      time1 = time.time()
       model.fit(x,y,batch_size = 1024,epochs=1)
      time2 = time.time()
     print(time2-time1)    ### This should not be this large ##



`

**Comment by gadagashwini** [comment]
@capilano I tried reproducing the issue with provided code but i received 
`name 'TPU_WORKER' is not defined`
Can you help us to reproduce the issue. Thanks!

**Comment by capilano** [comment]
TPU_WORKER is the TPU address. If you are using google colab (with TPU accelerator), I think you can leave it blank Just call the function without passing any argument. If that does not work,
 TPU_WORKER= 'grpc://{}'.format(os.environ['COLAB_TPU_ADDR'])

**Comment by gadagashwini** [comment]
@capilano I have tried reproducing the issue by adding piece of code `tf.keras.layers.GlobalAveragePooling2D()(x)` Since GlobalAveragePolling2D(x) was not defined. But still i am unable to replicate your issue. Provide us the full minimal code snippet. It will indeed help us to move faster.

**Comment by capilano** [comment]
Ok, I have changed the code. If you just copy/paste the code it should work. I have also changed the network, I am just using a few Conv layers to check, becausemaybe the Renet50 Model has some unsupported layers

**Comment by gadagashwini** [comment]
@capilano Thanks for the complete code. I am able to reproduce the issue now with Tf 1.14.0. Thanks!

**Comment by anki-xyz** [comment]
I'm facing the same issue. As ```fit_generator``` does not work, I am using the same strategy because of the large dataset and involved data augmentation. But calling ```fit``` is extremely slow. 
Things that come to my mind:
- re-loading the model to the TPU?
- data transfer?

Even when I load more data to decrease the ```fit``` calls (batch size > 2048), the sessions just fail. This makes me suspicious that it is maybe the data transfer...

**Comment by capilano** [comment]
You can use  create a tf records dataset and do this with fit

**Comment by jvishnuvardhan** [comment]
I could reproduce the issue. But I am not sure whether it is 100X slower or not. [Here](https://colab.sandbox.google.com/gist/jvishnuvardhan/50fceaf46f18b8a772a878c399f7676c/untitled570.ipynb) is the gist. Thanks!

**Comment by capilano** [comment]
@jvishnuvardhan Just to give you guys a heads up, one  can directly pass a dataset to model.fit and so multiple calls to fit are not really necessary if you are using a pipeline with only tensorflow functions for data augmentation.

**Comment by frankchn** [comment]
I am not surprised that the notebook is slow as the data processing is all happening on the Colab rather than the TPU system (which has much more processing power than the Colab VMs). 

With the notebook, the `data` variable contains 224 x 224 x 3 x 4 (bytes/float) x 1024 = 588 MB of data, which has to be transferred per step. Transferring this amount of data over the network to the TPU + encoding and decoding overhead would be non-trivial.

For performance reasons especially on non-trivial image models, you need to use tf.data Datasets with TF supported ops, and load the raw data from GCS.

**Comment by Santosh-Gupta** [comment]
Is `fit_generator` still not supported?
